executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,391 @@
+                        vars[*idx].unwrap(),
+                    );
+                }
+                InterpreterAction::WriteParam(idx, i) => {
+                    set_param(pparams, *i, vars[*idx].unwrap());
+                }
+                InterpreterAction::WriteKnown(c) => {
+                    set_known(
+                        self.first_column_id,
+                        self.column_count,
+                        known,
+                        params.row_offset,
+                        c.row_offset,
+                        c.id,
+                    );
+                }
+                InterpreterAction::MachineCall(result_vars, id, arguments) => {
+                    let mutable_state =
+                        unsafe { &*(params.mutable_state as *const MutableState<T, Q>) };
+
+                    let mut arg_values: Vec<_> = arguments
+                        .iter()
+                        .map(|a| match a {
+                            MachineCallArgument::Unknown(v) => (Some(v), Default::default()),
+                            MachineCallArgument::Known(v) => (
+                                None,
+                                self.evaluate_expression(&mut eval_stack, &vars, &v.to_rpn()),
+                            ),
+                        })
+                        .collect();
+
+                    // call machine
+                    let mut args = arguments
+                        .iter()
+                        .zip(arg_values.iter_mut())
+                        .map(|(a, v)| match a {
+                            MachineCallArgument::Unknown(_) => LookupCell::Output(&mut v.1),
+                            MachineCallArgument::Known(_) => LookupCell::Input(&v.1),
+                        })
+                        .collect::<Vec<_>>();
+
+                    mutable_state.call_direct(*id, &mut args[..]).unwrap();
+
+                    // write output to variables
+                    let mut var_idx = 0;
+                    args.into_iter().for_each(|arg| {
+                        if let LookupCell::Output(v) = arg {
+                            assert!(vars[result_vars[var_idx]].replace(*v).is_none());
+                            var_idx += 1;
+                        }
+                    });
+                }
+                InterpreterAction::Assertion(e1, e2, expected_equal) => {
+                    let lhs_value = self.evaluate_expression(&mut eval_stack, &vars, e1);
+                    let rhs_value = self.evaluate_expression(&mut eval_stack, &vars, e2);
+                    if *expected_equal {
+                        assert_eq!(lhs_value, rhs_value, "Assertion failed");
+                    } else {
+                        assert_ne!(lhs_value, rhs_value, "Assertion failed");
+                    }
+                }
+            }
+        }
+    }
+
+    // evaluates an expression using the provided variable values
+    fn evaluate_expression(
+        &self,
```

Comment:
`self` is unused as far as I can see. can this be a function on `RPNExpresssion`?